src/libstore/derivations.hh

```diff
@@ -77,6 +88,7 @@ enum struct DerivationType : uint8_t {
     DeferredInputAddressed,
     CAFixed,
     CAFloating,
+    Impure
 };
```

Comment:
I think it might be better to bite the bullet and do:
```c++
struct DerivationType_InputAddressed {
    bool deferred;
};

struct DerivationType_ContentAddressed {
    bool pure;
    bool fixed;
};

typedef std::variant<
    DerivationType_InputAddressed,
    DerivationType_ContentAddressed
> _DerivationTypeRaw

struct DerivationType : _DerivationTypeRaw {
    using Raw = _DerivationTypeRaw;
    using Raw::Raw;
    using InputAddressed = DerivationType_InputAddressed;
    using ContentAddressed = DerivationType_ContentAddressed;
};
```
This fills out the missing "pure, fixed" use-case. My belief is while "pure, fixed" may not be be very useful, implementing "missing cases" is good to ensure the code stays nice, and also test the code better.

This also explains why we don't need `DerivationOutputImpure`: Only the fixed/floating axis affects the sort of output it is.

Reply:
@edolstra would you mean if I took a stab at doing this, and then rebasing this PR getting rid of `DerivationOutputImpure` as outlined above?

I would like to do the cleanups first so the diff here is extra legible and easier to review.

Reply:
would "pure, fixed" be something like an enforced reproducibility check? or a cached CA outcome similar to builtins.storePath?

Either way, the "missing cases" viewpoint shows a good way to organize both the code and the underlying theory about how these types of derivations relate to each other.

Reply:
It's a fixed output derivation that has no special sandbox loopholes (e.g. it cannot access the network).

Glad you like the "missing cases" viewpoint too!